`not_empty_in` added to `codomain.py`

gxyd opened the pull request:
# TODO
- [x] Docstring added
- [x] function `not_empty_in` added
- [x] codomain for some non-rational functions now `not` made to raise `NotImplementedError`
- [x] TODO's added both to `codomain` and `not_empty_in`
- [x] https://github.com/sympy/sympy/pull/9779/files#r36607968

Commit:
`not_empty_in` added to `codomain.py`

Commit:
`not_empty_in` added to globals

Commit:
`tests` for `not_empty_in` added

Commit:
`docs` bug fixed

Commit:
ValueError for invalid `domain` in codomain

gxyd commented:
@hargup @aktech @jksuom please review.

gxyd commented on sympy/calculus/codomain.py:
```diff
@@ -102,6 +101,10 @@ def codomain_interval(f, set_val, *sym):
             local_minima = set([(-oo, True)])
 
+        if start_val.has(I) or end_val.has(I):
+                raise ValueError('Function does not Complement all points of %s '
```
I will change `Complement` to `contains`.

aktech commented on sympy/calculus/codomain.py:
```diff
@@ -151,3 +154,73 @@ def codomain_interval(f, set_val, *sym):
+
+
+def not_empty_in(fin_set, intrvl, *syms):
+    """ Finds the domain in which the Set input is not empty
```
`fin_set` --> `finite_set` and `intervl` --> `interval` or `sets` (Explicit is better than implicit)

The first argument of the following can only be `FiniteSet`, I guess, If so then you mean the following:

```
Finds the domain of the function(s) in `finite_set` in which the `finite_set` is not empty,
given it's range as `interval`.
```
or
```
Finds the domain of the function(s) in `finite_set` in which the `finite_set` is a subset of the `interval`.
```

aktech commented on sympy/calculus/codomain.py:
```diff
@@ -151,3 +154,73 @@ def codomain_interval(f, set_val, *sym):
+    y = Dummy('y')
+
+    def elm_domain(expr, *sym):
+        domain_union = S.EmptySet
```
You might want to write a couple of lines for the doc string. Even though if you make it private it's good to write a line about the functionality.

Commit:
variable names made more `explicit`

Commit:
`raised Error` and `typo fixed`

gxyd commented:
@aktech i have addressed your points, please have a look.

hargup commented on sympy/calculus/codomain.py:
```diff
@@ -45,6 +45,9 @@ def codomain(func, domain, *syms):
         raise ValueError('A Set must be given, not %s: %s' % (type(domain), domain))
 
+    # TODO: handle piecewise defined functions
+    # TODO: handle functions with infinitely many solutions (eg, sin, tan)
```
Transcendental functions in general.

hargup commented on sympy/calculus/codomain.py:
```diff
@@ -151,3 +154,83 @@ def codomain_interval(f, set_val, *sym):
                             else func.subs({symbol: i}) for i in domain])
+
+
+def not_empty_in(finite_set, sets, *syms):
```
This is not quite in the way we want it. Say you have an unevaluated intersection produced by `solveset`, for example `FiniteSet(x).intersect(Interval(0, 1))` for such sets the fact that the given set is empty or non-empty depends on the value of `x` and with your `not_empty_in` method we want to tell for what values of `x` this set will be non-empty. So, the input should be an unevaluated intersection and a symbol(s) and it should return another set.

Commit:
`not_empty_in` signature changed

gxyd commented:
@hargup I have made the changes.

hargup commented on sympy/calculus/codomain.py:
```diff
@@ -151,3 +154,78 @@ def codomain_interval(f, set_val, *sym):
                             else func.subs({symbol: i}) for i in domain])
+
+
+def not_empty_in(finSet_intersection, *syms):
```
Please don't use CamelCase, by convention they are for class names, use lowercase_with_underscores.

hargup commented on sympy/calculus/codomain.py:
```diff
@@ -151,3 +154,78 @@ def codomain_interval(f, set_val, *sym):
+        if isinstance(invert_expr, Intersection):
+            invert_set = invert_expr.args[1]
+        elif isinstance(invert_expr, Complement):
+            invert_set = invert_expr.args[0].args[1]
```
What is `args[0].args[1]` for Complement, does it always exist?

gxyd replied:
> What is args[0].args[1]

```
>>> a = solveset_real(x/(x - 1) - y, x)
Intersection((-oo, oo), {y/(y - 1)}) \ {1}
```

Hence the solution returned by `solveset_real` can be an instance of `Complement`
`a.args[0].args[1]` gives `FiniteSet(y/(y - 1))`

> does it always exist

Yes, i guess for when `solveset_real` returns an instance of `Complement`.

hargup replied:
> 
Yes, i guess for when `solveset_real` returns an instance of `Complement`.

That's true only if the first argument of `Complement` is intersection. We can have Complements like: `Interval(-oo, oo) \ {y}`

gxyd replied:
I agree that it can be the case.
There are two reasons why i have not taken care of such a case
1. First i can not find an example where `solveset_real(f - y, x)` returns` Interval(-oo, oo) - FiniteSet(y)` or something similar.

2. Secondly currently we have `Interval - FiniteSet` returns `EmptySet` which has not been fixed.

Though adding this particular case would not be tough. So do you think we can take care this particular case in the `future` ?

hargup replied:
Yes, you should take care of such cases. I don't see anything which is preventing the code from generating `Complement` where the first parameter is not an intersection.

hargup replied:
btw you can't just ignore the second parameter of `Complement` for example, in:
```python
In [27]: solveset_real((x**2 - 3*x + 2)/(x-1) - y, x)
Out[27]: Intersection((-oo, oo), {y + 2}) \ {1}
```
`y+2` cannot take the value `1` so you have to remove either `y=-1` from the domain or `{1}` from the codomain, the second option seems easier.

Commit:
`camel-case` variable changed and one more test added

gxyd commented:
I have added a new `test` in the latest commit which addresses `.args[0].args[1]`. Though a `test` for this was previously added but marked `@XFAIL`.

hargup commented on sympy/calculus/codomain.py:
```diff
@@ -102,6 +101,10 @@ def codomain_interval(f, set_val, *sym):
         elif start_val is S.NegativeInfinity or end_val is S.NegativeInfinity:
             local_minima = set([(-oo, True)])
 
+        if start_val.has(I) or end_val.has(I):
```
better alternative: `if (not start_val.is_real) or (not end_val.is_real)`:

gxyd replied:
Yes i realized that later on. I am gonna change that in new commit.

hargup commented on sympy/calculus/codomain.py:
```diff
@@ -151,3 +154,78 @@ def codomain_interval(f, set_val, *sym):
+        domain_union = S.EmptySet
+
+        # find the inverse of items in the finite_set
+        invert_expr = solveset_real(expr - y, symbol)
```
It is recommended that you use `solveset` with domain=`S.Reals` instead of `solveset_real`.

Commit:
`domain=S.Reals` in `solveset` and elements out of `codomain` removed

Commit:
`not_empty_in` removed from `__init__`

Commit:
`changes` to import

gxyd commented:
@hargup i have made the changes, thanks. 
But i still disagree that `Complement(Interval, FiniteSet)` is possible as result of `solveset` since of the reason that: we are solving that in `S.Reals` so i think the first argument of Complement is always going to be Intersection(S.Reals, FiniteSet) but still. If you can give me a contradictory example. Then i can look into that problem.

hargup commented:
Nope it is not necessary, if we have the right assumptions on the variables then `Intersection(S.Reals, FiniteSet)` can turn out to be a `FiniteSet
```python
In [12]: x = Symbol('x',real=True)

In [13]: FiniteSet(x).intersect(S.Reals)
Out[13]: {x}
```

Second you are making a fundamental error of deduction, just because you or anyone  cannot find an example of something doesn't imply that an example cannot exist. As Carl Sagan famously said:

> **absence of evidence is not evidence of absence**

General statements needs general proofs and contradictory examples can only be used to disprove them. As a crude example if you claim that all men have beard, then you cannot say that it is true because you have never seen a man who doesn't have beard. You have to show that not having a beard is somehow contradictory with the definition of man. On the other hand if I have to disprove you then I just need to show you a man without natural beard.

Commit:
Complement(FiniteSet, FiniteSet) tests added

Commit:
ValueError for not-specifying symbol

gxyd commented:
@hargup  thanks for the clarification. I have your points into consideration. Please have a look.

gxyd commented:
@hargup I mostly to leave space for worst case where certain conditions may be met.
> absence of evidence is not evidence of absence

I hope that below answer is the evidence. Though `i have added the code for what you said`.

> Nope it is not necessary, if we have the right assumptions on the variables then Intersection(S.Reals, FiniteSet) can turn out to be a `FiniteSet
In [12]: x = Symbol('x',real=True)
In [13]: FiniteSet(x).intersect(S.Reals)
Out[13]: {x}

Actually even if the user applies restrictions on the `x` being real, that does not make a difference of `solveset` returns the solution.
Consider this as an example:
```
>>> x = Symbol('x', real=True)
>>> a = FiniteSet(2*x).intersect(S.Reals)
>>> a
{2*x}
>>> not_empty_in(a, x)
(-oo, oo)
```

This is the answer:
Here when i want to find the `inverse of 2*x` (after defining `x` is real) then i use solveset(expr - y) where `y = Dummy('y')` now since i  have not put restrictions on `y` being real. `Intersection(S.Reals, FiniteSet)` is never going to be `reduced` to `FiniteSet`.

gxyd commented on sympy/calculus/codomain.py:
```diff
@@ -151,3 +154,98 @@ def codomain_interval(f, set_val, *sym):
+    else:
+        raise NotImplementedError('more than one variables %s not handled' % (syms,))
+
+    y = Dummy('y')
```
This is the `Dummy variable`, `y`

hargup commented on sympy/calculus/codomain.py:
```diff
@@ -151,3 +154,98 @@ def codomain_interval(f, set_val, *sym):
+    ValueError
+        The input is not valid.
+    RuntimeError
+        It is a bug, please report to the github issue tracker.
```
you can provide a link to the issue tracker here.

hargup commented on sympy/calculus/codomain.py:
```diff
@@ -151,3 +154,98 @@ def codomain_interval(f, set_val, *sym):
+    if not isinstance(finite_set, FiniteSet):
+        raise ValueError('A FiniteSet must be given, not %s: %s' % (type(finite_set), finite_set))
+    if len(syms) == 1:
+        symbol = syms[0]
```
`symbol` is a module in SymPy, you shouldn't override it. `sym = syms[0]` should work.

hargup commented:
> Here when i want to find the `inverse of 2*x` (after defining `x` is real) then i use solveset(expr - y) where `y = Dummy('y')` now since i  have not put restrictions on `y` being real. `Intersection(S.Reals, FiniteSet)` is never going to be `reduced` to `FiniteSet`.

Thats how the `solveset` code behave currently and its behaviour can change in future, for example I was considering the we replace all the symbols in the equations with the real assumption as keeping `domain=S.Reals` means that we are operating only on reals. Will we make this change or not is not the question here, the point is that your code should be stable to small changes in the behaviour of ohter part of the code. Your code shouldn't especially depend on the fringe behaviours which are not mentioned in the documentation.

hargup commented:
@gxyd can you give the coverage report?

gxyd commented:
@hargup  Here is the [coverage-report](https://medium.com/@gauravdhingra/coverage-report-for-not-empty-in-function-in-sympy-3f714cca4cd5)

Commit:
names changed and link added

gxyd commented:
@hargup can you look why it is failing on `pypy`. I am looking for how to make the `@XFAIL` test pass.(i think some problem in `codomain` is there`)

Commit:
bug in `tests` fixed

Commit:
approch for elm_domain changed

gxyd commented:
I have now changed my approach for the function. The reason i have changed:

1. It's more basic, or quite easy to understand.
2. It is fast compared to the previous approach (why fast ? since i am using just one time `solveset` with no use of `codomain`)
3. It solves the problem that `inverse` of an expression can have multiple values, increasing the number of expressions for which `codomain` has to be found.

How i came to realise this approach: While thinking about the problem with the previously `@XFAIl` ed test, i talked to my friend about this. So found that i this approach is more basic and good.

Here is the link to the [new-coverage-report](https://medium.com/@gauravdhingra/new-not-empty-in-function-in-sympy-coverage-report-c926b107bd0d) .
Let me know you think.

@hargup @aktech

Commit:
bug fixed

gxyd commented on sympy/calculus/codomain.py:
```diff
@@ -143,11 +147,118 @@ def codomain_interval(f, set_val, *sym):
+        """ Finds the domain of an expression in any given interval """
+
+        _start = min(intrvl.start, intrvl.end)
+        _end = max(intrvl.start, intrvl.end)
```
There probably is no need for this `max-min` thing. I will change this.

Commit:
typos fixed

Commit:
names changed

gxyd commented:
It is ready for the final review

hargup commented on sympy/calculus/codomain.py:
```diff
@@ -143,11 +147,119 @@ def codomain_interval(f, set_val, *sym):
+
+        if intrvl.right_open:
+            if _end is S.Infinity:
+                _sing1 = solveset(expr.as_numer_denom()[1], symb, domain=S.Reals)
```
Why aren't you using `_sing` defined above?

hargup replied:
Also please use full names, `_singularities` instead of `_sing`.

hargup commented on sympy/calculus/codomain.py:
```diff
@@ -143,11 +147,119 @@ def codomain_interval(f, set_val, *sym):
+
+        if intrvl.left_open:
+            if _start is S.NegativeInfinity:
+                _sing2 = solveset(expr.as_numer_denom()[1], symb, domain=S.Reals)
```
Same as https://github.com/sympy/sympy/pull/9779/files#r37316809

hargup commented:
@gxyd you also need to add this function to `__init__.py` file. Let me know when you are done.

Commit:
un-necessary declarations removed

gxyd commented:
@hargup done. I have removed those un-necessary `_sing1` and `_sing2`, changed names to `_singularities`. Added to `__init__.py` .

hargup commented:
+1, please add a reference to this PR in the release notes.

gxyd commented:
Thanks Harsh for all the points regarding this PR.
